pkg/ccl/kvccl/kvfollowerreadsccl/followerreads.go

```diff
@@ -203,6 +205,49 @@ func (o *followerReadOracle) useClosestOracle(
 // whether a query can be used with a follower read.
 var followerReadOraclePolicy = replicaoracle.RegisterPolicy(newFollowerReadOracle)
 
+type bulkOracle struct {
+	cfg       replicaoracle.Config
+	locFilter roachpb.Locality
+}
+
+var _ replicaoracle.Oracle = bulkOracle{}
+
+// NewBulkOracle returns an oracle for planning bulk operations, which will plan
+// balancing random across all replicas if follower reads are enabled.
+func NewBulkOracle(cfg replicaoracle.Config, locFilter roachpb.Locality) replicaoracle.Oracle {
+	return bulkOracle{cfg: cfg, locFilter: locFilter}
+}
+
+// ChoosePreferredReplica implements the replicaoracle.Oracle interface.
+func (r bulkOracle) ChoosePreferredReplica(
+	ctx context.Context,
+	_ *kv.Txn,
+	desc *roachpb.RangeDescriptor,
+	leaseholder *roachpb.ReplicaDescriptor,
+	_ roachpb.RangeClosedTimestampPolicy,
+	_ replicaoracle.QueryState,
+) (_ roachpb.ReplicaDescriptor, ignoreMisplannedRanges bool, _ error) {
+	if leaseholder != nil && !checkFollowerReadsEnabled(uuid.UUID{} /*not used*/, r.cfg.Settings) {
+		return *leaseholder, false, nil
+	}
+	replicas, err := kvcoord.NewReplicaSlice(ctx, r.cfg.NodeDescs, desc, nil, kvcoord.AllExtantReplicas)
+	if err != nil {
+		return roachpb.ReplicaDescriptor{}, false, sqlerrors.NewRangeUnavailableError(desc.RangeID, err)
+	}
+	if r.locFilter.NonEmpty() {
+		var matches []roachpb.ReplicaDescriptor
```

Comment:
I suppose? but like... this is job planning -- compared to what we'll do _after_ planning this is peanuts. But sure, as long as this struct doesn't need to be threadsafe, if we're worried about perf here, I can reuse this slice, and also make it `[]int` of offsets into the replica slice to reduce its size.

Reply:
actually no, I don't think I should worry about this allocating here. Right above this we're allocating a new slice every time in `kvcoord.NewReplicaSlice` so until this shows up in a profile I'm calling this premature optimization (and one which would make it non-thread safe).

Reply:
the problem with this is that it will not show up in this functions profile; it will show up as gc runtime...
But we do know that this function will be called thousands of times -- why allocate more than we need to?
A quick micro benchmark will certainly show that.

Reply:
The existing oracle -- which we're calling thousands of times just like this one -- calls `replicaSliceOrErr` for every span which allocates a whole ` kvcoord.NewReplicaSlice` every call. If we want to start golfing allocations in this thing -- including the existing implementations as well -- then fine we can do that, but let's do it in an isolated performance-change-only PR, that has its own backport discussion, not in a behavior change.

Reply:
Sounds good.  Memoizing allocations does not make sense when you're doing one off queries; it's just seeing what
trying to plan a flow w/ 2.5M ranges would do, that made me suggest it.  Totally fine in a standalone PR.